### PR TITLE
CuArrays -> CUDA

### PR DESCRIPTION
**Pull request by DhairyaLGandhi**
This updates the compatibility bounds on CuArrays and moves it to CUDA.jl. CUDA.jl seems to lower bound on Julia 1.4 (@maleadt any chance we could run it on 1.3 somehow?).

### PR Checklist

- [x] Tests are added
- [x] Entry in NEWS.md
- [ ] Documentation, if applicable
- [ ] Final review from `@MikeInnes` or `@dhairyagandhi96` (for API changes).

### CODE REVIEW

**Comment by dhairyagandhi96** [comment]
bors try

**Comment by cossio** [comment]
fixes https://github.com/FluxML/Flux.jl/issues/1194

**Comment by cossio** [comment]
Why do we need to support Julia 1.3?

**Comment by dhairyagandhi96** [comment]
We try to maintain backwards compatibility, for older versions and lts versions of the language. The 1.0 support wasn't dropped too long back.

**Comment by dhairyagandhi96** [comment]
I'm a bit confused, does this fix #1194 or is that a blocker?

**Comment by cossio** [comment]
This PR fixes https://github.com/FluxML/Flux.jl/issues/1194, as long as it makes Flux compatible with CUDA. I thought my comment would automatically close the issue when this gets merged, but apparently only the PR author can make these auto-close comments.

**Comment by dhairyagandhi96** [comment]
bors try

**Comment by dhairyagandhi96** [comment]
bors try

**Comment by dhairyagandhi96** [comment]
bors try

**Comment by dhairyagandhi96** [comment]
How about that! Needs https://github.com/FluxML/Zygote.jl/pull/673

**Comment by dhairyagandhi96** [comment]
I believe CUDA.jl would support 1.3 as well, so we don't have to up the version bounds on Flux. As suggested by Tim, I'll move over the tests on `CUDA#master`

File: Project.toml [hunk 1]
```diff
@@ -30,7 +30,7 @@ AbstractTrees = "0.2, 0.3"
 Adapt = "1, 2.0"
 CodecZlib = "0.5, 0.6, 0.7"
 Colors = "0.8, 0.9, 0.10, 0.11, 0.12"
-CuArrays = "2"
+CUDA = "0"
```
Comment (lassepe) [inline Project.toml:33]: Why 0? Shouldn't this be 1?

**Comment by cossio** [comment]
Now that https://github.com/FluxML/Zygote.jl/pull/701 is merged, can we get a tagged release of Flux with this?

**Comment by CarloLucibello** [comment]
@cossio we also have a Zygote 0.4 tag fom yesterday which is essentially equivalent to 0.5 but without the new CUDA.jl support, so you should be able to use Flux with the new features.
A new flux release may take some time since we still need to merge a few breaking changes

File: Project.toml (hunk 1 above)
Comment (dhairyagandhi96) [reply Project.toml:33]: Since we can support CUDA@0.1 on 1.4, but now that 1.0 is out, I will update this.

**Comment by dhairyagandhi96** [comment]
Seems like something is going off testing with GPUs and CUDA@1.

The run for CUDA@0.1 is passing though

bors try

**Comment by CarloLucibello** [comment]
you should bump Zygote compat to 0.5 and NNlib to 0.7

**Comment by dhairyagandhi96** [comment]
I did try that, that lead to other tests failing

File: docs/src/gpu.md [hunk 2]
```diff
@@ -1,17 +1,17 @@
...
 ## GPU Usage
 
-Support for array operations on other hardware backends, like GPUs, is provided by external packages like [CuArrays](https://github.com/JuliaGPU/CuArrays.jl). Flux is agnostic to array types, so we simply need to move model weights and data to the GPU and Flux will handle it.
+Support for array operations on other hardware backends, like GPUs, is provided by external packages like [CUDA](https://github.com/JuliaGPU/CUDA.jl). Flux is agnostic to array types, so we simply need to move model weights and data to the GPU and Flux will handle it.
 
-For example, we can use `CuArrays` (with the `cu` converter) to run our [basic example](models/basics.md) on an NVIDIA GPU.
+For example, we can use `CUDA` (with the `cu` converter) to run our [basic example](models/basics.md) on an NVIDIA GPU.
 
-(Note that you need to have CUDA available to use CuArrays – please see the [CuArrays.jl](https://github.com/JuliaGPU/CuArrays.jl) instructions for more details.)
+(Note that you need to have CUDA available to use CUDA – please see the [CUDA.jl](https://github.com/JuliaGPU/CUDA.jl) instructions for more details.)
```
Comment (CarloLucibello) [inline docs/src/gpu.md:11]:
```suggestion
(Note that you need to have CUDA available to use CUDA.CuArrays – please see the [CUDA.jl](https://github.com/JuliaGPU/CUDA.jl) instructions for more details.)
```

File: docs/src/gpu.md [hunk 3]
```diff
@@ -1,17 +1,17 @@
...
-NVIDIA GPU support should work out of the box on systems with CUDA and CUDNN installed. For more details see the [CuArrays](https://github.com/JuliaGPU/CuArrays.jl) readme.
+NVIDIA GPU support should work out of the box on systems with CUDA and CUDNN installed. For more details see the [CUDA](https://github.com/JuliaGPU/CUDA.jl) readme.
 
 ## GPU Usage
 
-Support for array operations on other hardware backends, like GPUs, is provided by external packages like [CuArrays](https://github.com/JuliaGPU/CuArrays.jl). Flux is agnostic to array types, so we simply need to move model weights and data to the GPU and Flux will handle it.
+Support for array operations on other hardware backends, like GPUs, is provided by external packages like [CUDA](https://github.com/JuliaGPU/CUDA.jl). Flux is agnostic to array types, so we simply need to move model weights and data to the GPU and Flux will handle it.
 
-For example, we can use `CuArrays` (with the `cu` converter) to run our [basic example](models/basics.md) on an NVIDIA GPU.
+For example, we can use `CUDA` (with the `cu` converter) to run our [basic example](models/basics.md) on an NVIDIA GPU.
```
Comment (CarloLucibello) [inline docs/src/gpu.md:9]:
```suggestion
For example, we can use `CUDA.CuArrays` (with the `cu` converter) to run our [basic example](models/basics.md) on an NVIDIA GPU.
```

**Comment by CarloLucibello** [comment]
> I did try that, that lead to other tests failing

We have to fix whatever comes up, Zygote < 0.5 doesn't support CUDA.jl, so we have to drop support in flux

**Comment by dhairyagandhi96** [comment]
Fair enough I will update the bounds

**Comment by dhairyagandhi96** [comment]
bors try

**Comment by dhairyagandhi96** [comment]
I am noticing this locally, which suggests NNPACK might be reading garbage 
```julia
CNN: Test Failed at /home/dhairyagandhi96/Flux.jl/test/layers/conv.jl:38
  Expression: sum(op) === 0.0f0
   Evaluated: -2.6233066f31 === 0.0f0
Stacktrace:
 [1] top-level scope at /home/dhairyagandhi96/Flux.jl/test/layers/conv.jl:38
 [2] top-level scope at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Test/src/Test.jl:1107
 [3] top-level scope at /home/dhairyagandhi96/Flux.jl/test/layers/conv.jl:18
CNN: Test Failed at /home/dhairyagandhi96/Flux.jl/test/layers/conv.jl:56
  Expression: Flux.mse(bias(ip), op) ≈ 4.0f0
   Evaluated: NaN32 ≈ 4.0f0
Stacktrace:
 [1] top-level scope at /home/dhairyagandhi96/Flux.jl/test/layers/conv.jl:56
 [2] top-level scope at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/Test/src/Test.jl:1107
 [3] top-level scope at /home/dhairyagandhi96/Flux.jl/test/layers/conv.jl:18
```

**Comment by CarloLucibello** [comment]
One of the errors is #1244, should be fixed by https://github.com/FluxML/Zygote.jl/pull/706

**Comment by DhairyaLGandhi** [comment]
bors try

**Comment by DhairyaLGandhi** [comment]
bors try

**Comment by CarloLucibello** [comment]
some of the failures are due to #1245 , we should fix it in zygote , although it's not clear when we introduced the regression

**Comment by CarloLucibello** [comment]
This warning
https://travis-ci.org/github/FluxML/Flux.jl/jobs/700819230#L360
 exposes some mixed types problems that were previously handled by NNlib through conversion. I think this should be fixed on the Flux side, probably just finding where in tests we are using Float64 and change it to Float32

**Comment by CarloLucibello** [comment]
This error instead is weird https://travis-ci.org/github/FluxML/Flux.jl/jobs/700819230#L339
Maybe we can just deactivate NNPACK in NNlib and reconsider later

**Comment by DhairyaLGandhi** [comment]
Yeah, that's what we had done with the environment variable `NNLIB_USE_NNPACK`

**Comment by DhairyaLGandhi** [comment]
Why would handling it on Flux be beneficial?

**Comment by DhairyaLGandhi** [comment]
bors try

**Comment by DhairyaLGandhi** [comment]
Looking closer into the failings, it seems like the error is with the reported gradients, which worked fine until Zygote@0.4.20

```julia
julia> cuback(1.f0)
(Base.RefValue{Any}((cell = Base.RefValue{Any}((Wi = Float32[-0.028762825 -0.06328906 … -0.003087131 -0.046078723; -0.009768781 -0.021495003 … -0.0010484891 -0.015649816; … ; -0.013579503 -0.029880026 … -0.0014574961 -0.021754682; -0.010848033 -0.023869764 … -0.0011643259 -0.017378805], Wh = Float32[0.0 0.0 … 0.0 0.0; 0.0 0.0 … 0.0 0.0; … ; 0.0 0.0 … 0.0 0.0; 0.0 0.0 … 0.0 0.0], b = Float32[-0.06590673, -0.022384048, 0.069970295, -0.03814131, -0.019469338, 0.0, 0.0, 0.0, 0.0, 0.0, 0.26342788, 0.26144245, 0.09559605, 0.1319688, 0.32197332, -0.05778172, -0.019014964, 0.044791963, -0.03111588, -0.024857031], h = nothing, c = nothing)), init = nothing, state = (Float32[-0.08846992, 0.07109468, -0.15598866, -0.12647739, -0.002516523], Float32[0.48143172, 0.38970852, 0.38037091, 0.29649684, 0.40764716]))), Float32[-0.104312934; -0.112346165; … ; 0.091621324; -0.23611495])
``` 

and don't Zygote@0.4.21 onwards

```julia
julia> cuback(1.f0)
(Base.RefValue{Any}((cell = nothing, init = nothing, state = (Float32[-0.04327198, 0.024098, -0.077963844, -0.0560417, -0.017322173], Float32[0.46127376, 0.26323906, 0.18762967, 0.38597986, 0.39259586]))), Float32[0.17995863; 0.12015681; … ; 0.021471716; 0.031247895])
```

notice the `nothing` in the gradient for `cell`.

**Comment by CarloLucibello** [comment]
yes it's #1245 . Maybe we should get this merged marking those tests as broken and fix somewhere else

**Comment by DhairyaLGandhi** [comment]
bors try

**Comment by DhairyaLGandhi** [comment]
I think we can merge this now, we need to patch NNlib and check the regression in Zygote.

**Pull request closed without merging**